test_gpt2.cu

```diff
@@ -136,72 +155,55 @@ int main(int argc, char *argv[]) {
+            cudaMemcpy(calculated_grads.qkvw, model.grads.qkvw, L * 3*C * C * sizeof(float), cudaMemcpyDeviceToHost);
+            cudaMemcpy(calculated_grads.qkvb, model.grads.qkvb, L * 3*C * sizeof(float), cudaMemcpyDeviceToHost);
+            cudaMemcpy(calculated_grads.ln1w, model.grads.ln1w, L * C * sizeof(float), cudaMemcpyDeviceToHost);
+            cudaMemcpy(calculated_grads.ln1b, model.grads.ln1b, L * C * sizeof(float), cudaMemcpyDeviceToHost);
+            cudaMemcpy(calculated_grads.wte, model.grads.wte, V * C * sizeof(float), cudaMemcpyDeviceToHost);
+            cudaMemcpy(calculated_grads.wpe, model.grads.wpe, maxT * C * sizeof(float), cudaMemcpyDeviceToHost);
+            int gradoks[16];
+            gradoks[0] =  check_tensor(calculated_grads.lnfb, expected_grads.lnfb, C, "lnfb", 1.3e-06f);
+            gradoks[1] =  check_tensor(calculated_grads.lnfw, expected_grads.lnfw, C, "lnfw", 3.1e-06f);
+            gradoks[2] =  check_tensor(calculated_grads.fcprojw, expected_grads.fcprojw, L * C * 4*C, "fcprojw", 6.8e-06f);
+            gradoks[3] =  check_tensor(calculated_grads.fcprojb, expected_grads.fcprojb, L * C, "fcprojb", 1.9e-06f);
+            gradoks[4] =  check_tensor(calculated_grads.fcw, expected_grads.fcw, L * 4*C * C, "fcw", 1.3e-05f);
+            gradoks[5] =  check_tensor(calculated_grads.fcb, expected_grads.fcb, L * 4*C, "fcb", 4.8e-06f);
+            gradoks[6] =  check_tensor(calculated_grads.ln2w, expected_grads.ln2w, L * C, "ln2w", 1.2e-04f);
+            gradoks[7] =  check_tensor(calculated_grads.ln2b, expected_grads.ln2b, L * C, "ln2b", 1e-05f);
+            gradoks[8] =  check_tensor(calculated_grads.attprojw, expected_grads.attprojw, L * C * C, "attprojw", 4e-06f);
+            gradoks[9] =  check_tensor(calculated_grads.attprojb, expected_grads.attprojb, L * C, "attprojb", 2.9e-06f);
+            gradoks[10] = check_tensor(calculated_grads.qkvw, expected_grads.qkvw, L * 3*C * C, "qkvw", 2.1e-05f);
+            gradoks[11] = check_tensor(calculated_grads.qkvb, expected_grads.qkvb, L * 3*C, "qkvb", 5.5e-06f);
+            gradoks[12] = check_tensor(calculated_grads.ln1w, expected_grads.ln1w, L * C, "ln1w", 9e-05f);
+            gradoks[13] = check_tensor(calculated_grads.ln1b, expected_grads.ln1b, L * C, "ln1b", 2.1e-05f);
+            gradoks[14] = check_tensor(calculated_grads.wte, expected_grads.wte, V * C, "wte", 3.8e-05f);
+            gradoks[15] = check_tensor(calculated_grads.wpe, expected_grads.wpe, maxT * C, "wpe", 1.7e-06);
+            for (int i = 0; i < 16; i++) {
+                allok = allok & gradoks[i];
+            }
+            // compare the gradients on the parameters all at once
+            cudaMemcpy(calculated_grads_memory, model.grads_memory, model.num_parameters * sizeof(float), cudaMemcpyDeviceToHost);
+            if (!check_tensor(calculated_grads_memory, expected_grads_memory, model.num_parameters, "all grads", 2e-04f)) {
+                allok = 0;
```

Comment:
I need to put this code back in.  Merge missed it.

char* src_iterator = (char*)calculated_grads_memory;
            float* dst_iterator = (float*)converted_grads_memory;
            for (size_t i = 0; i < NUM_PARAMETER_TENSORS; i++) {
                if (model.param_sizeof[i] == sizeof(float)) {
                    memcpy(dst_iterator, src_iterator, model.param_elements[i] * sizeof(float));
                } else {
                    assert(model.param_sizeof[i] == sizeof(floatX));
                    for (size_t j = 0; j < model.param_elements[i]; j++) {
                        dst_iterator[j] = ((floatX*)src_iterator)[j];
                    }
                }
                src_iterator += model.param_elements[i] * model.param_sizeof[i];
                dst_iterator += model.param_elements[i];
            }

Reply:
Pretty bad diff with **ENABLE_BF16**

LOGITS - OK w/ tol = 2.300000e+01, maxdiff=1.384476e+01
LOSS - OK w/ tol = 2.300000e+01, maxdiff=2.971172e-03
**grads - NOT OK w/ tol = 1.000000e-02, faults = 37639, maxdiff=4.316021e+08**
grads[36] 0.0659188 -0.0063978, diff=7.231519e-02
grads[314] -0.0466318 -0.0173328, diff=2.929912e-02
grads[374] 0.0013208 0.0227368, diff=2.141586e-02
grads[430] 0.0155038 -0.0796288, diff=9.513110e-02
grads[496] 0.1215828 -0.1077288, diff=2.293102e-01
grads[1198] 0.0952158 0.0776668, diff=1.754867e-02